README.md
# This is a forked version, since the original auther not maintain anymore.

Below is the original readme.
-----

# [Like cVim? Consider donating! https://www.paypal.me/1995eaton](https://www.paypal.me/1995eaton)

# What is cVim?

Vim for Google Chrome. I hate using the mouse, especially after learning Vim. With my desktop (Linux), I have a lot of key bindings that make doing things easier: I open Chrome with `Alt+w`, I close a window with `Alt+Shift+d`, I open a terminal with `Alt+t`. This is harder to do with Chrome because it has no section for customizing keyboard shortcuts, and it is still necessary to use the mouse to do things like click links. cVim aims to eliminate this problem as best as the Chrome extensions API will allow it to.

# Where can I get cVim?

 * There are two ways:
  * You can install it through the [Chrome web store](https://chrome.google.com/webstore/detail/cvim/ihlenndgcmojhcghmfjfneahoeklbjjh)
  * You can download the `.zip` file [here](https://github.com/1995eaton/chromium-vim/archive/master.zip) and enable cVim by going to the `chrome://extensions` URL and checking developer mode, then pointing Chrome to the unzipped folder via the `Load unpacked extensions...` button.

# Why is this different than Vimium, ViChrome, or Vrome?

These extensions do a wonderful job of adding Vim-like keybindings to Google Chrome, but they lack many of the features that Firefox Addon, Pentadactyl, have.

 * What features does cVim add to Chrome?
  * Google/IMDB/Wikipedia/Amazon/Duckduckgo/Yahoo/Bing search completion
  * Support for custom search engines
  * History and Bookmark search/completion with bookmark folder support
  * Caret/Visual mode
  * Efficient link hints (with support for custom mappings)
  * Support for custom keyboard mappings
  * Regex page search with highlighting
  * Command bar with tab-completion
  * Smooth scrolling

# cVim Help
### cVimrc

 * Boolean cVimrc settings are enabled with the command ```'set' + <SETTING_NAME>``` and disabled with
   the command ```'set' + no<SETTING_NAME>``` (for example, ```set regexp``` and ```set noregexp```)
 * Boolean cVimrc settings can be inversed by adding "!" to the end
 * Other settings are defined with ```=``` used as a separator and are prefixed by ```let``` (for example, ```let hintcharacters="abc"```)

| setting                             | type                               | description                                                                               | default                                                                     |
| ----------------------------------- | ---------------------------------- | ----------------------------------------------------------------------------------------- | --------------------------------------------------------------------------: |
| searchlimit                         | integer                            | set the amount of results displayed in the command bar                                    | 25                                                                          |
| scrollstep                          | integer                            | set the amount of pixels scrolled when using the scrollUp and scrollDown commands         | 70                                                                          |
| timeoutlen                          | integer                            | The amount of time to wait for a `<Leader>` mapping in milliseconds                       | 1000                                                                        |
| fullpagescrollpercent               | integer                            | set the percent of the page to be scrolled by when using the scrollFullPageUp and scrollFullPageDown commands | 0                                                       |
| typelinkhintsdelay                  | integer                            | the amount of time (in milliseconds) to wait before taking input after opening a link hint with typelinkhints and numerichints enabled | 300                            |
| scrollduration                      | integer                            | the duration of smooth scrolling                                                          | 500                                                                         |
| vimport                             | integer                            | set the port to be used with the `editWithVim` insert mode command                        | 8001                                                                        |
| zoomfactor                          | integer / double                   | the step size when zooming the page in/out                                                | 0.1                                                                         |
| scalehints                          | boolean                            | animate link hints as they appear                                                         | false                                                                       |
| hud                                 | boolean                            | show the heads-up-display                                                                 | true                                                                        |
| regexp                              | boolean                            | use regexp in find mode                                                                   | true                                                                        |
| ignorecase                          | boolean                            | ignore search case in find mode                                                           | true                                                                        |
| linkanimations                      | boolean                            | show fade effect when link hints open and close                                           | false                                                                       |
| numerichints                        | boolean                            | use numbers for link hints instead of a set of characters                                 | false                                                                       |
| dimhintcharacters                   | boolean                            | dim letter matches in hint characters rather than remove them from the hint               | true                                                                        |
| defaultnewtabpage                   | boolean                            | use the default chrome://newtab page instead of a blank page                              | false                                                                       |
| cncpcompletion                      | boolean                            | use `<C-n>` and `<C-p>` to cycle through completion results (requires you to set the nextCompletionResult keybinding in the chrome://extensions page (bottom right) | false |
| smartcase                           | boolean                            | case-insensitive find mode searches except when input contains a capital letter           | true                                                                        |
| incsearch                           | boolean                            | begin auto-highlighting find mode matches when input length is greater thant two characters | true                                                                      |
| typelinkhints                       | boolean                            | (numerichints required) type text in the link to narrow down numeric hints                | false                                                                       |
| autohidecursor                      | boolean                            | hide the mouse cursor when scrolling (useful for Linux, which doesn't auto-hide the cursor on keydown) | false                                                          |
| autofocus                           | boolean                            | allows websites to automatically focus an input box when they are first loaded            | true                                                                        |
| insertmappings                      | boolean                            | use insert mappings to navigate the cursor in text boxes (see bindings below)             | true                                                                        |
| smoothscroll                        | boolean                            | use smooth scrolling                                                                      | false                                                                       |
| autoupdategist                      | boolean                            | if a GitHub Gist is used to sync settings, pull updates every hour (and when Chrome restarts)   | false                                                                 |
| nativelinkorder                     | boolean                            | Open new tabs like Chrome does rather than next to the currently opened tab               | false                                                                       |
| showtabindices                      | boolean                            | Display the tab index in the tab's title                                                  | false                                                                       |
| sortlinkhints                       | boolean                            | Sort link hint lettering by the link's distance from the top-left corner of the page      | false                                                                       |
| localconfig                         | boolean                            | Read the cVimrc config from `configpath` (when this is set, you connot save from cVim's options page | false                                                            |
| completeonopen                      | boolean                            | Automatically show a list of command completions when the command bar is opened           | false                                                                       |
| configpath                          | string                             | Read the cVimrc from this local file when configpath is set                               | ""                                                                          |
| changelog                           | boolean                            | Auto open the changelog when cVim is updated                                              | true                                                                        |
| completionengines                   | array of strings                   | use only the specified search engines                                                     | ["google", "duckduckgo", "wikipedia", "amazon"]                             |
| blacklists                          | array of strings                   | disable cVim on the sites matching one of the patterns                                    | []                                                                          |
| mapleader                           | string                             | The default `<Leader>` key                                                                | \                                                                           |
| defaultengine                       | string                             | set the default search engine                                                             | "google"                                                                    |
| locale                              | string                             | set the locale of the site being completed/searched on (see example configuration below)  | ""                                                                          |
| homedirectory                       | string                             | the directory to replace `~` when using the `file` command                                | ""                                                                          |
| qmark &lt;alphanumeric charcter&gt; | string                             | add a persistent QuickMark (e.g. ```let qmark a = ["http://google.com", "http://reddit.com"]```) | none                                                                 |
| previousmatchpattern                | string (regexp)                    | the pattern looked for when navigating a page's back button                               | ((?!last)(prev(ious)?&#124;newer&#124;back&#124;«&#124;less&#124;&lt;&#124;‹&#124; )+) |
| nextmatchpattern                    | string (regexp)                    | the pattern looked for when navigation a page's next button                               | ((?!first)(next&#124;older&#124;more&#124;&gt;&#124;›&#124;»&#124;forward&#124; )+)    |
| hintcharacters                      | string (alphanumeric)              | set the default characters to be used in link hint mode                                   | "asdfgqwertzxcvb"                                                           |
| barposition                         | string ["top", "bottom"]           | set the default position of the command bar                                               | "top"                                                                       |
| langmap                             | string                             | set a list of characters to be remapped (see vims langmap)                                | ""                                                                          |

### Example configuration
```vim
" Settings
set nohud
set nosmoothscroll
set noautofocus " The opposite of autofocus; this setting stops
                " sites from focusing on an input box when they load
set typelinkhints
let searchlimit = 30
let scrollstep = 70
let barposition = "bottom"

let locale = "uk" " Current choices are 'jp' and 'uk'. This allows cVim to use sites like google.co.uk
                  " or google.co.jp to search rather than google.com. Support is currently limited.
                  " Let me know if you need a different locale for one of the completion/search engines
let hintcharacters = "abc123"

let searchengine dogpile = "http://www.dogpile.com/search/web?q=%s" " If you leave out the '%s' at the end of the URL,
                                                                    " your query will be appended to the link.
                                                                    " Otherwise, your query will replace the '%s'.

" This will do the same thing as above, except typing ':tabnew withbase' into to command bar
" without any search parameters will open 'http://www.dogpile.com'
let searchengine withbase = ["http://www.dogpile.com", "http://www.dogpile.com/search/web?q=%s"]

" alias ':g' to ':tabnew google'
command g tabnew google

let completionengines = ["google", "amazon", "imdb", "dogpile"]

let searchalias g = "google" " Create a shortcut for search engines.
                             " For example, typing ':tabnew g example'
                             " would act the same way as ':tabnew google example'

" Open all of these in a tab with `gnb` or open one of these with <N>goa where <N>
let qmark a = ["http://www.reddit.com", "http://www.google.com", "http://twitter.com"]

let blacklists = ["https://mail.google.com/*", "*://mail.google.com/*", "@https://mail.google.com/mail/*"]
" blacklists prefixed by '@' act as a whitelist

let mapleader = ","

" Mappings

map <Leader>r reloadTabUncached
map <Leader>x :restore<Space>

" This remaps the default 'j' mapping
map j scrollUp

" You can use <Space>, which is interpreted as a
" literal " " character, to enter buffer completion mode
map gb :buffer<Space>

" This unmaps the default 'k' mapping
unmap k

" This unmaps the default 'h', 'j', 'k', and 'l' mappings
unmap h j k l

" This remaps the default 'f' mapping to the current 'F' mapping
map f F

" Toggle the current HUD display value
map <C-h> :set hud!<CR>

" Switch between alphabetical hint characters and numeric hints
map <C-i> :set numerichints!<CR>

map <C-u> rootFrame
map <M-h> previousTab
map <C-d> scrollPageDown
map <C-e> scrollPageUp
iunmap <C-y>
imap <C-m> deleteWord

" Create a variable that can be used/referenced in the command bar
let @@reddit_prog = 'http://www.reddit.com/r/programming'
let @@top_all = 'top?sort=top&t=all'
let @@top_day = 'top?sort=top&t=day'

" TA binding opens 'http://www.reddit.com/r/programming/top?sort=top&t=all' in a new tab
map TA :tabnew @@reddit_prog/@@top_all<CR>
map TD :tabnew @@reddit_prog/@@top_day<CR>

" Use paste buffer in mappings
map T :tabnew wikipedia @"<CR>

" Code blocks (see below for more info)
getIP() -> {{
httpRequest({url: 'http://api.ipify.org/?format=json', json: true},
            function(res) { Status.setMessage('IP: ' + res.ip); });
}}
" Displays your public IP address in the status bar
map ci :call getIP<CR>

" Script hints
echo(link) -> {{
  alert(link.href);
}}
map <C-f> createScriptHint(echo)

let configpath = '/path/to/your/.cvimrc'
set localconfig " Update settings via a local file (and the `:source` command) rather
                " than the default options page in chrome
" As long as localconfig is set in the .cvimrc file. cVim will continue to read
" settings from there
```

### Blacklists
 * The blacklists setting uses a custom inplementation of Chrome's @match pattern guidelines.
   See https://developer.chrome.com/extensions/match_patterns for a description of the syntax.


### Site-specific Configuration
 * You can enable certain rc settings for sites using the blacklist match pattern as described above
```vim
" this will enable the config block below on the domain 'reddit.com'
site '*://*.reddit.com/*' {
      unmap j
      unmap k
      set numerichints
}
```

### Running commands when a page loads
 * In a similar fashion to the site-specific configuration described above, cVim can run commands when a page is loaded with the `call` keyword
```vim
" In this case, when pages with a file ending in '.js' are loaded,
" cVim will pin the tab and then scroll down
site '*://*/*.js' {
      call :pintab
      call scrollDown
}
```

### Mappings
 * Normal mappings are defined with the following structure: ```map <KEY> <MAPPING_NAME>```
 * Insert mappings use the same structure, but use the command "imap" instead of "map"
 * Control, meta, and alt can be used also:
```vim
<C-u> " Ctrl + u
<M-u> " Meta + u
<A-u> " Alt  + u
```
 * It is also possible to unmap default bindings with ```unmap <KEY>``` and insert bindings with ```iunmap <KEY>```
 * To unmap all default keybindings, use ```unmapAll```. To unmap all default insert bindings, use ```iunmapAll```

### Tabs
 * Commands that open links (`:tabnew` and `:open`) have three different properties
  * `!` => Open in a new tab
  * `$` => Open in a new window
  * `|` => Open in an incognito window
  * `&` => Open in a new tab (inactive/unfocused)
  * `*` => Pin the tab
  * `?` => Treat the query as a search
  * `=` => Treat the query as a URL
 * The use of these properties are best explained with examples:

```vim
:open! google<CR> " This is the same as :tabnew google<CR>

:open google!<CR> " This is another way of writing the above
                  " (these flags can can be added to either
                  " the base command or the end of the final command)

:open& google<CR> " This will open Google in a new inactive tab

:open$ google<CR> " This will open Google in a new window

:open&* google<CR> " The will open Google in a new inactive, pinned tab

:tabnew google&*<CR> " Once again, this will do the same thing as the above command

:open google&*<CR> " Again, same as above

:open google!& " Here, the & flag will cancel out the ! flag,
               " opening Google in a new inactive tab

" More examples
:bookmarks my_bookmark.com&  " inactive,new tab
:bookmarks&* my_bookmark.com " inactive,pinned,new tab
:bookmarks! my_bookmark.com  " new tab
:bookmarks$ my_bookmark.com  " new window
:bookmarks my_bookmark.com   " same tab
```

### Code blocks
 * Code blocks allow you to interact with cVim's content scripts via the cVimrc.
 * Since code blocks use `eval(...)`, you should only use them if you know what you're doing.

```JavaScript
" To be used by the code block
set hintset_a

" Create a code block named switchHintCharacters
switchHintCharacters -> {{
  // We are now in JavaScript mode

  // Settings are contained in an object named settings
  settings.hintset_a = !settings.hintset_a;
  if (settings.hintset_a) {
    settings.hintcharacters = 'abc'; // equivalent to "let hintcharacters = 'abc'"
  } else {
    settings.hintcharacters = 'xyz';
  }

  // Propagate the current settings to all tabs for the
  // rest of the session
  PORT('syncSettings', { settings: settings });

  // Display cVim's status bar for 2 seconds.
  Status.setMessage('Hint Set: ' + (true ? 'a' : 'b'), 2);
}}

" Run the JavaScript block
map <Tab> :call switchHintCharacters<CR>
```

### Completion Engines
  * These are a list of completion engines that can be used in the command bar. They can be set
    by assigning their names to an array with the `completionengines` variable.
    * google, wikipedia, youtube, imdb, amazon, google-maps, wolframalpha, google-image, ebay,
      webster, wictionary, urbandictionary, duckduckgo, answers, google-trends, google-finance,
      yahoo, bing, themoviedb
  * Example usage:
```vim

CHANGELOG.md
# cVim Changelog

# [Like cVim? Consider donating! https://www.paypal.me/1995eaton](https://www.paypal.me/1995eaton)

## 1.3.1 (Unreleased)

- Will fix the console "Unchecked runtime.lastError: The message port closed before a response was received." issue

## 1.3.0 (2020-04-22)

- Fix the message hint not working issue
- Fix insert cursor in command mode

## 1.2.99 (2017-12-17)

- Updated list of `chrome://` completion URLs when `:chrome` is used
- Fixed issue with `openPasteTab` [#573](https://github.com/1995eaton/chromium-vim/issues/573)
- Allow global site matches in cVimrc. For example, the following will unmap `j` on all sites:

```
site '*://*' {
  unmap j
}
```

- Fix issue with certain hints not showing up (like the GitHub profile icon in the top-right corner of the page).

## 1.2.98 (2017-10-13)

- Fix [#569](https://github.com/1995eaton/chromium-vim/issues/569) (mappings like `map <C-A-M-i> nextTab` would work, but `map <C-M-A-i> nextTab` wouldn't)
- Let `@"` represent the paste buffer in command bar commands (e.g. `tabnew wikipedia "@` would search Wikipedia for whatever is in the paste buffer)

## 1.2.97 (2017-09-20)

- Make `^` go to beginning of line when in insert mode

## 1.2.96 (2017-08-31)

- Fix vulnerability where sites are able to trigger keyboard events that would be interpreted by cVim ([#572](https://github.com/1995eaton/chromium-vim/issues/572))

## 1.2.95 (2017-08-16)

- Redesign options page and fixed CSS issues for smaller screens

## 1.2.94 (2017-08-13)

- Add dates to changelog

## 1.2.93 (2017-07-19)

- Fix exception being thrown on React.js input forms

## 1.2.92 (2017-07-06)

- Fix corrupted zip file issue with last release

## 1.2.91

- [Implement jumplist feature](https://github.com/1995eaton/chromium-vim/issues/553) (`previousScrollPosition` is `<C-o>` and `nextScrollPosition` is `<C-i>`).

## 1.2.90 (2017-06-29)

- Re-implement editWithVim

## 1.2.89 (2017-06-20)

- If you are currently running an old version of the [cvim_server.py](https://github.com/1995eaton/chromium-vim/blob/master/cvim_server.py) script (keeping it up to date locally), please update to the latest version (in the link above)! For the time being, the `editWithVim` feature has been disabled due to a security concern.
- Bug fixes

## 1.2.88

- Fix issue with `:buffer` command

## 1.2.87 (2017-03-23)

- Fix input box issues with new versions of Chrome

## 1.2.86 (2017-02-17)

- Add `openQuickMarkWindowed` command

## 1.2.85 (2016-06-22)

- Bug fixes

## 1.2.84 (2016-04-22)

- Bug fixes

## 1.2.83 (2016-04-19)

- Fix bug in Chrome 50+

## 1.2.82 (2016-04-02)

- Fix `:new`/`:winopen` bug
- Fix scrolling bug when experimental JavaScript features are enabled
- Fix several insert mapping bugs
- Allow multiple unmaps using a single statement (e.g. `unmap h j k l`)
- Added new window modifier for `:tabnew` and related commands

## 1.2.81 (2015-12-20)

- Smooth scrolling is now disabled by default (add `set smoothscroll` to your settings to get it back)
- Bug fixes

## 1.2.80 (2015-10-06)

- Scrolling improvements
- Bug fixes

## 1.2.79 (2015-07-09)

- Content scripts will now be reloaded if they lose connection to the
  background scripts (e.g. restarting the extension)

## 1.2.78

- Default completion engines are now only `["google", "duckduckgo", "wikipedia", "amazon"]`.
  The other engines still exist, but you'll need to enable them with the `completionengines` option.
  - See [mappings.html#completion-engines](./mappings.html#completion-engines) for more info.
- Added "themoviedb" to completion engines

## 1.2.77 (2015-06-09)

- Bug fixes
- Add support for local ".cvimrc" files

## 1.2.76 (2015-05-01)

- Fixed issue with buffer switching. Partially breaks fix from last version

## 1.2.75 (2015-04-29)

- Fixed issue where cVim sometimes stops responding after opening the command bar

## 1.2.74 (2015-04-07)

- Bug fixes
- Added `sortlinkhints` setting

## 1.2.73 (2015-02-28)

- Link hint / command bar redesign (may need to reset CSS for this to take effect)
- Improved autocompletion matching
- Several bugfixes

## 1.2.72 (2015-02-12)

- Fix an issue with command bar z-index on YouTube: [#237](https://github.com/1995eaton/chromium-vim/issues/237)
- Fix an issue with tab completion with the `:buffer` command: [#238](https://github.com/1995eaton/chromium-vim/issues/238)

## 1.2.71

- Fix incompatibilities with some non-American keyboard layouts

## 1.2.70

- Fix a cncpcompletion bug
- Added the `createScriptHint` command
- Multiple bugfixes

## 1.2.69

- Fix a bug with text areas ([#231](https://github.com/1995eaton/chromium-vim/issues/231))

## 1.2.68

- Bug fixes from iframe addition in 1.2.68
- Added code block feature

## 1.2.67

- Moved command bar to separate frame (issue [#85](https://github.com/1995eaton/chromium-vim/issues/85)).
- Several security patches

## 1.2.66

- Added `nativelinkeorder`, `vimcommand`, and `vimport` options
- `cvim_socket.py` is now `cvim_server.py`
- Bug fixes

## 1.2.65

- Better smooth scrolling performance
- Fix a URI encoding bug with search engines

## 1.2.64

- Improved cVimrc parsing (with [PEG.js](http://pegjs.majda.cz/))
- The options page will now report errors on which line has improper syntax.
  You can check Chrome's JavaScript console for more info on what the error was
  and why it occurred.
- Better history searching
- Several bug fixes

## 1.2.63

- Fix for Google search and the `:` key (and some others)

## 1.2.62

- Added autoupdategist option
- Added lastUsedTab command
- The goToSource command is now `:viewsource&<CR>`
- Removed the mapping blacklist feature for the blacklists array (e.g. `let blacklists = ["*://*.reddit.com/* <Up> <Down>])
- The syntax for an extended version of this feature can now be found [here](https://github.com/1995eaton/chromium-vim#site-specific-configuration)

## 1.2.61

- Minor bug fixes from last update
- Fix issue [#120](https://github.com/1995eaton/chromium-vim/issues/120)

## 1.2.60


background_scripts/options.js
var storageMethod = 'local',
    settings = {},
    Options = {};

var defaultSettings = {
  searchlimit: 25,
  scrollstep: 70,
  fullpagescrollpercent: 0,
  typelinkhintsdelay: 300,
  qmarks: {},
  sites: {},
  searchengines: {},
  searchaliases: {},
  hud: true,
  regexp: true,
  scalehints: false,
  linkanimations: false,
  sortlinkhints: false,
  ignorecase: true,
  numerichints: false,
  cncpcompletion: false,
  smartcase: true,
  incsearch: true,
  autohidecursor: false,
  typelinkhints: false,
  autofocus: true,
  insertmappings: true,
  defaultnewtabpage: false,
  dimhintcharacters: true,
  smoothscroll: false,
  autoupdategist: false,
  nativelinkorder: false,
  showtabindices: false,
  changelog: true,
  localconfig: false,
  completeonopen: false,
  debugcss: false, // Always use default COMMANDBARCSS
  scrollduration: 500,
  zoomfactor: 0.10,
  zoomLevelAllTabs: 1.6,
  configpath: '',
  locale: '',
  mapleader: '\\',
  timeoutlen: 1000,
  defaultengine: 'google',
  hintcharacters: 'asdfgqwertzxcvb',
  homedirectory: '',
  langmap: '',
  completionengines: ['google', 'duckduckgo', 'wikipedia', 'amazon'],
  nextmatchpattern: '((?!first)(next|older|more|>|›|»|forward| )+)',
  previousmatchpattern: '((?!last)(prev(ious)?|newer|back|«|less|<|‹| )+)',
  barposition: 'top',
  vimport: 8001,
  blacklists: [],
  RC: '',
  MAPPINGS: '',
  GISTURL: '',
  FUNCTIONS: {},
  COMMANDBARCSS: '#cVim-command-bar, #cVim-command-bar-mode, #cVim-command-bar-input, #cVim-command-bar-search-results,\n.cVim-completion-item, .cVim-completion-item .cVim-full, .cVim-completion-item .cVim-left,\n.cVim-completion-item .cVim-right {\n  font-family: Helvetica, Helvetica Neue, Neue, sans-serif, monospace, Arial;\n  font-size: 10pt !important;\n  -webkit-font-smoothing: antialiased !important;\n}\n\n#cVim-command-bar {\n  position: fixed;\n  z-index: 2147483646;\n  background-color: #1b1d1e;\n  color: #bbb;\n  display: none;\n  box-sizing: content-box;\n  box-shadow: 0 3px 3px rgba(0,0,0,0.4);\n  left: 0;\n  width: 100%;\n  height: 20px;\n}\n\n#cVim-command-bar-mode {\n  display: inline-block;\n  vertical-align: middle;\n  box-sizing: border-box;\n  padding-left: 2px;\n  height: 100%;\n  width: 10px;\n  padding-top: 2px;\n  color: #888;\n}\n\n#cVim-command-bar-input {\n  background-color: #1b1d1e;\n  color: #bbb;\n  height: 100%;\n  right: 0;\n  top: 0;\n  width: calc(100% - 10px);\n  position: absolute;\n}\n\n#cVim-command-bar-search-results {\n  position: fixed;\n  width: 100%;\n  overflow: hidden;\n  z-index: 2147483647;\n  left: 0;\n  box-shadow: 0 3px 3px rgba(0,0,0,0.4);\n  background-color: #1c1c1c;\n}\n\n.cVim-completion-item, .cVim-completion-item .cVim-full, .cVim-completion-item .cVim-left, .cVim-completion-item .cVim-right {\n  text-overflow: ellipsis;\n  padding: 1px;\n  display: inline-block;\n  box-sizing: border-box;\n  vertical-align: middle;\n  overflow: hidden;\n  white-space: nowrap;\n}\n\n.cVim-completion-item:nth-child(even) {\n  background-color: #1f1f1f;\n}\n\n.cVim-completion-item {\n  width: 100%; left: 0;\n  color: #bcbcbc;\n}\n\n.cVim-completion-item[active] {\n  width: 100%; left: 0;\n  color: #1b1d1e;\n  background-color: #f1f1f1;\n}\n\n.cVim-completion-item[active] span {\n  color: #1b1d1e;\n}\n\n.cVim-completion-item .cVim-left {\n  color: #fff;\n  width: 37%;\n}\n\n.cVim-completion-item .cVim-right {\n  font-style: italic;\n  color: #888;\n  width: 57%;\n}\n\n\n#cVim-link-container, .cVim-link-hint,\n#cVim-hud, #cVim-status-bar {\n  font-family: Helvetica, Helvetica Neue, Neue, sans-serif, monospace, Arial;\n  font-size: 10pt !important;\n  -webkit-font-smoothing: antialiased !important;\n}\n\n#cVim-link-container {\n  position: absolute;\n  pointer-events: none;\n  width: 100%; left: 0;\n  height: 100%; top: 0;\n  z-index: 2147483647;\n}\n\n.cVim-link-hint {\n  position: absolute;\n  color: #302505 !important;\n  background-color: #ffd76e !important;\n  border-radius: 2px !important;\n  padding: 2px !important;\n  font-size: 8pt !important;\n  font-weight: 500 !important;\n  text-transform: uppercase !important;\n  border: 1px solid #ad810c;\n  display: inline-block !important;\n  vertical-align: middle !important;\n  text-align: center !important;\n  box-shadow: 2px 2px 1px rgba(0,0,0,0.25) !important;\n}\n\n.cVim-link-hint_match {\n  color: #777;\n  text-transform: uppercase !important;\n}\n\n\n#cVim-hud {\n  background-color: rgba(28,28,28,0.9);\n  position: fixed !important;\n  transition: right 0.2s ease-out;\n  z-index: 24724289;\n}\n\n#cVim-hud span {\n  padding: 2px;\n  padding-left: 4px;\n  padding-right: 4px;\n  color: #8f8f8f;\n  font-size: 10pt;\n}\n\n#cVim-frames-outline {\n  position: fixed;\n  width: 100%;\n  height: 100%;\n  left: 0;\n  top: 0;\n  right: 0;\n  z-index: 9999999999;\n  box-sizing: border-box;\n  border: 3px solid yellow;\n}\n'
};

chrome.storage.onChanged.addListener(function(changes) {
  if (!changes.hasOwnProperty('sessions')) {
    settings = changes.settings ? changes.settings.newValue : defaultSettings;
  }
});

Options.refreshSettings = function(callback) {
  for (var key in defaultSettings) {
    if (settings[key] === void 0) {
      settings[key] = defaultSettings[key];
    }
  }
  if (callback) {
    callback();
  }
};

Options.saveSettings = function(request) {
  settings = request.settings;
  for (var key in settings.qmarks) {
    Quickmarks[key] = settings.qmarks[key];
  }
  this.refreshSettings(function() {
    chrome.storage[storageMethod].set({settings: settings});
    if (request.sendSettings) {
      Options.sendSettings();
    }
  });
};

Options.sendSettings = function() {
  activePorts.forEach(function(port) {
    port.postMessage({
      type: 'sendSettings',
      settings: settings
    });
  });
};

Options.getSettings = function(request, sender) {
  this.refreshSettings(function() {
    chrome.tabs.sendMessage(sender.tab.id, {
      action: 'sendSettings',
      settings: request.reset ? defaultSettings : settings
    });
  });
};

Options.setDefaults = function() {
  settings = defaultSettings;
  this.saveSettings(settings);
};

Options.getDefaults = function(request, sender, callback) {
  callback(defaultSettings);
};

// retrieve default and current settings
Options.getAllSettings = function(request, sender, callback) {
  callback({
    defaults: defaultSettings,
    current: settings
  });
};

Options.updateBlacklistsMappings = function() {
  var rc = Utils.compressArray(settings.RC.split(/\n+/)),
      i, index, line;
  if (settings.BLACKLISTS) {
    settings.blacklists = settings.BLACKLISTS.split(/\n+/);
    delete settings.BLACKLISTS;
  }
  for (i = 0; i < rc.length; ++i) {
    if (/ *let *blacklists *= */.test(rc[i])) {
      rc.splice(i, 1);
      index = i;
    }
  }
  settings.blacklists = Utils.uniqueElements(settings.blacklists);
  if (settings.blacklists.length) {
    line = 'let blacklists = ' + JSON.stringify(settings.blacklists);
    if (index) {
      rc = rc.slice(0, index).concat(line).concat(rc.slice(index));
    } else {
      rc.push(line);
    }
  }
  settings.RC = rc.join('\n');
  Options.saveSettings({settings: settings});
};

Options.fetchGist = function() {
  httpRequest({
    url: settings.GISTURL + (settings.GISTURL.indexOf('raw') === -1 &&
             settings.GISTURL.indexOf('github') !== -1 ? '/raw' : '')
  }).then(function(res) {
    var updated;
    try {
      updated = RCParser.parse(res);
    } catch (e) {
      console.error('cVim Error: error parsing config file');
    }
    updated.GISTURL = settings.GISTURL;
    updated.COMMANDBARCSS = settings.COMMANDBARCSS;
    Options.saveSettings({
      settings: updated,
      sendSettings: true
    });
    if (updated.autoupdategist) {
      window.setTimeout(Options.fetchGist, 1000 * 60 * 60);
    }
  });
};

chrome.storage[storageMethod].get('settings', function(data) {
  if (data.settings) {
    settings = data.settings;
    Quickmarks = settings.qmarks;
  }
  if (settings.debugcss)
    settings.COMMANDBARCSS = defaultSettings.COMMANDBARCSS;
  this.refreshSettings();
  this.updateBlacklistsMappings();
  if (settings.autoupdategist && settings.GISTURL) {
    this.fetchGist();
  }
}.bind(Options));

chrome.runtime.onMessage.addListener(function(request, sender, callback) {
  if (Options.hasOwnProperty(request.action)) {
    Options[request.action](request, sender, callback);
  }
});
